--[[
	ğŸ’– Thanks for using Satchel ğŸ’–

	Satchel is a modern open-source alternative to Roblox's default backpack ğŸ’

	ğŸ› Playground: https://www.roblox.com/games/13592168150
	ğŸ“° DevForum: https://devforum.roblox.com/t/2451549
	ğŸ“‚ GitHub: https://github.com/RyanLua/Satchel

	ğŸ”§ Installation:
		1. Move the Satchel module script into StarterPlayerScripts
		2. Enable AutoUpdate in the Satchel module script package link
		3. Edit the Satchel module script attributes to customize the UI
		4. Read the documentation to learn how to use Satchel using the above links
]]
